src/compute/src/memory_management/policy.rs

```diff
@@ -0,0 +1,301 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::atomic::AtomicU64;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use pretty_bytes::converter::convert;
+use risingwave_batch::task::BatchManager;
+use risingwave_common::error::Result;
+use risingwave_common::util::epoch::Epoch;
+use risingwave_stream::task::LocalStreamManager;
+
+/// `MemoryControlStats` contains the necessary information for memory control, including both batch
+/// and streaming.
+pub struct MemoryControlStats {
+    pub batch_memory_usage: usize,
+    pub streaming_memory_usage: usize,
+    pub jemalloc_allocated_mib: usize,
+    pub lru_watermark_step: u64,
+    pub lru_watermark_time_ms: u64,
+    pub lru_physical_now_ms: u64,
+}
+
+pub type MemoryControlPolicy = Box<dyn MemoryControl>;
+
+pub trait MemoryControl: Send + Sync {
+    fn apply(
+        &self,
+        total_compute_memory_bytes: usize,
+        barrier_interval_ms: u32,
+        prev_memory_stats: MemoryControlStats,
+        batch_manager: Arc<BatchManager>,
+        stream_manager: Arc<LocalStreamManager>,
+        watermark_epoch: Arc<AtomicU64>,
+    ) -> MemoryControlStats;
+
+    fn describe(&self, total_compute_memory_bytes: usize) -> String;
```

Comment:
Is it possible to pass the unchange parts, e.g. `total_compute_memory_bytes`, `barrier_interval_ms` into concrete memory control policy implementation so that we can simplify API? For example, this way we don't need the `describe` method but implement `Display` trait would be enough.

Reply:
IMO the policy should only contain fields regarding the policy itself while excluding any information about user configurations like the CN memory.